tests/test_pyglmnet.py

```diff
@@ -164,61 +164,91 @@ def test_glmnet():
-                assert_allclose(first_beta, second_beta, rtol=0.05, atol=1e-2)
+        for reg_lambda in reg_lambdas:
+            betas_ = list()
+            for solver in solvers:
+
+                np.random.seed(random_state)
+
+                X_train = np.random.normal(0.0, 1.0, [n_samples, n_features])
+                y_train = simulate_glm(distr, beta0, beta, X_train,
+                                       sample=False)
+                alpha = 0.5
+                loss_trace = list()
+
+                def callback(beta):
+                    Tau = None
+                    eta = 2.0
+                    group = None
+
+                    loss_trace.append(
+                        _loss(distr, alpha, Tau, reg_lambda,
+                              X_train, y_train, eta, group, beta))
+
+                glm = GLM(distr, learning_rate=learning_rate,
+                          reg_lambda=reg_lambda, tol=1e-3, max_iter=5000,
+                          alpha=alpha, solver=solver,
+                          score_metric=score_metric,
+                          random_state=random_state, callback=callback)
+                assert(repr(glm))
+
+                glm.fit(X_train, y_train)
+
+                # verify loss decreases or increases only slightly
+                loss_trace_decreased = np.diff(loss_trace) <= 1e-3
+                if np.any(~loss_trace_decreased):
+                    idx_of_first_increase = 2 + \
+                        np.min(np.nonzero(~loss_trace_decreased))
+                assert np.all(loss_trace_decreased), \
+                    ('Loss increased between iterations'
+                     ' on distr={d} solver={s} with'
+                     ' reg_lambda={rl}'
+                     '\n Loss trace:\n {lt}'
+                     ).format(d=distr, s=solver,
+                              rl=reg_lambda,
+                              lt=loss_trace[:idx_of_first_increase])
+
+                if reg_lambda == 0.0:
+                    # check that the true model can be recreated
+                    # almost perfectly when no regularization is applied
+                    # verify loss at convergence = loss when beta=beta_
+                    l_true = _loss(distr, alpha, np.eye(beta.shape[0]),
+                                   reg_lambda, X_train, y_train, 2.0, None,
+                                   np.concatenate(([beta0], beta)))
+                    assert_allclose(loss_trace[-1], l_true,
+                                    rtol=1e-4, atol=1e-5,
+                                    err_msg=('Final loss trace value different'
+                                             ' from true loss '
+                                             ' on distr={d} solver={s}'
+                                             ' with reg_lambda={rl}'
+                                             ).format(d=distr, s=solver,
+                                                      rl=reg_lambda))
```

Comment:
Those are there because it made it a ton easier for me to solve this problem. I'd recommend keeping them so that the problematic distr/solver is immediately clear without manual pdb work.